### PR TITLE
Proxy user avatar images from gravatar

### PR DESCRIPTION
**Pull request by segiddins**
See https://bundler.slack.com/archives/C02F27TKQ/p1704394288403519?thread_ts=1703757212.084219&cid=C02F27TKQ for discussion

Closes https://github.com/rubygems/rubygems.org/issues/4318

Original suggestion: https://github.com/rubygems/rubygems.org/pull/4104#pullrequestreview-1653384916

### CODE REVIEW

**Review by indirect — APPROVED** [review]
this looks great, thank you! hopefully our avatars will be a little bit less confusing and identical now.

File: app/controllers/avatars_controller.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
+class AvatarsController < ApplicationController
+  before_action :find_user
+  before_action :set_size
+  before_action :set_theme
+
+  def show
+    gravatar_url = @user.gravatar_url(size: @size, default: "404", secure: true)
+
+    resp = Faraday.get(gravatar_url, nil,
```
Comment (colby-swandale) [inline app/controllers/avatars_controller.rb:9]: What's Faraday's timeout here? If Gravatar's servers are being particularly spotty, we wouldn't want to be waiting around forever for an answer. Should we put a timeout of something like 5 seconds in?

**Review by colby-swandale — APPROVED** [review]
So glad we can get the avatar images back 👏🏻

File: app/controllers/avatars_controller.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      send_data(resp.body, type: resp.headers["content-type"], disposition: "inline", filename:)
+    elsif resp.status == 404
+      fastly_expires_in(5.minutes)
+
+      # means gravatar doesn't have an avatar for this user
+      # we'll just redirect to our default avatar instead, so everything is cachable
+      redirect_to default_avatar_url
+    else
+      # any other error, just redirect to our default avatar
+      # this includes 404s, 500s, etc
```
Comment (martinemde) [inline app/controllers/avatars_controller.rb:27]: Nit: Not 404s because of above.

**Review by martinemde — APPROVED** [review]
Awesome!

**Pull request merged**